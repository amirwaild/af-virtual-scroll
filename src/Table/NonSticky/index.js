import React, { memo, useMemo, Fragment } from "react";
import PropTypes from "prop-types";
import { css, cx } from "emotion";

import useApi from "../../useApi";
import TableHeader from "./TableHeader";
import TableFooter from "./TableFooter";

import Thead from "../common/Thead";
import Tfoot from "../common/Tfoot";

import ColgroupCached from "../common/ColgroupCached";
import TbodyScrollerCached from "../common/TbodyScrollerCached";
import ScrollContainer from "../common/ScrollContainer";
import Tbody from "../common/Tbody";


/*
    border-box is important, because head th widths are synced with td widths
*/
const wrapperClass = css`
    display: flex;
    flex-flow: column nowrap;
`;

const headerFooterClass = css`
    flex: 0 0 auto;
    table-layout: fixed;
    min-width: 100%;
`;

const hiddenHeaderFooterClass = css`
    visibility: hidden;
    th, td {
        line-height: 0 !important;
        padding-top: 0 !important;
        padding-bottom: 0 !important;
        border-top: none !important;
        border-bottom: none !important;
    }
`;

const subscribeEvents = [
    "headless-mode-changed",
    "totals-changed"
];

const TableHeaderCached = <TableHeader className={headerFooterClass} />;
const TableFooterCached = <TableFooter className={headerFooterClass} />;

const NonSticky = ({
    className,
    tbodyRef,
    scrollContainerRef,
    getRowExtraProps,
    RowComponent,
    CellComponent,
    fixedLayout,
    ...props
}) => {

    const { headlessMode, totals } = useApi( subscribeEvents );

    return (
        <div className={cx(wrapperClass, className )} {...props}>
            {headlessMode?null:TableHeaderCached}
            <ScrollContainer ref={scrollContainerRef} fixedLayout={fixedLayout}>
                {useMemo(() => (
                    <Fragment>
                        {ColgroupCached}
                        <Thead className={hiddenHeaderFooterClass} />
                        <Tfoot className={hiddenHeaderFooterClass} />
                        {TbodyScrollerCached}
                        <Tbody
                            tbodyRef={tbodyRef}
                            getRowExtraProps={getRowExtraProps}
                            RowComponent={RowComponent}
                            CellComponent={CellComponent}
                        />
                    </Fragment>
                ), [ fixedLayout, getRowExtraProps, RowComponent, CellComponent ])}
            </ScrollContainer>
            {totals&&TableFooterCached}
        </div>
    )
}

export default memo( NonSticky );