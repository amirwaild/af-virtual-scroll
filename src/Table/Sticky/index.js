import React, { memo, useMemo, useEffect } from "react";
import { css, cx } from "emotion";

import Colgroup from "../common/Colgroup";
import TbodyScroller from "../common/TbodyScroller";
import ScrollContainer from "../../common/ScrollContainer";
import Thead from "../common/Thead";
import Tbody from "../common/Tbody";
import Tfoot from "../common/Tfoot";
import BodyTable from "../common/BodyTable";
import useApi from "../../useApi";

const theadTfootCommonClass = css`
    td, th {
        position: sticky;
        z-index: 1;
    }
`;

const theadClass = cx( theadTfootCommonClass, css`
    td, th {
        top: 0;
    }
`);

const tfootClass = cx( theadTfootCommonClass, css`
    td, th {
        bottom: 0;
    }
`);

const SUBSCRIBE_EVENTS = [
    "#headlessMode",
    "#totals"
];


/*
    Todo:
        measure thead & tfoot heights in order to properly calculate available space for rows
*/


const Sticky = ({
    tbodyRef,
    scrollContainerRef,
    getRowExtraProps,
    getCellExtraProps,
    RowComponent,
    CellComponent,
    TotalsCellComponent,
    fixedLayout,
    ...props
}) => {

    const { headlessMode, totals } = useApi( SUBSCRIBE_EVENTS );

    if( process.env.NODE_ENV !== "production" ){
        /*
            https://bugs.chromium.org/p/chromium/issues/detail?id=702927
        */

        const areTotalsPresent = totals && totals.length !== 0;

        useEffect(() => {
            if( !headlessMode || areTotalsPresent ){

                const table = scrollContainerRef.current.querySelector( "table" );
                const tableStyle = getComputedStyle( table );

                if( tableStyle.borderCollapse === "collapse" ){
                    console.warn(
                        "You use sticky table version. Due to special border behavior when scrolling, use border-collpase: separate.%o",
                        "https://bugs.chromium.org/p/chromium/issues/detail?id=702927"
                    );
                }
            }
        }, [ headlessMode, areTotalsPresent ]);
    }
    
    return (
        <ScrollContainer ref={scrollContainerRef} reportScrollLeft {...props}>
            {useMemo(() => (
                <BodyTable fixedLayout={fixedLayout}>
                    <Colgroup />
                    {headlessMode?null:<Thead className={theadClass} />}
                    <TbodyScroller />
                    <Tbody
                        tbodyRef={tbodyRef}
                        getRowExtraProps={getRowExtraProps}
                        getCellExtraProps={getCellExtraProps}
                        RowComponent={RowComponent}
                        CellComponent={CellComponent}
                    />
                    {totals && (
                        <Tfoot
                            className={tfootClass}
                            TotalsCellComponent={TotalsCellComponent}
                        />
                    )}
                </BodyTable>
            ), [ headlessMode, fixedLayout, totals, getRowExtraProps, getCellExtraProps, RowComponent, CellComponent, TotalsCellComponent ])}
        </ScrollContainer>
    );
}

export default memo( Sticky );