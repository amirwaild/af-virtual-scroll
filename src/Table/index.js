import React, { memo, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import isPositionStickySupported from "../utils/isPositionStickySupported";
import Context from "../Context";
import VirtualTableDataStore from "../models/Table";

import RowComponentDefault from "./common/Row";
import CellComponentDefault from "./common/Cell";
import RowCountWarningContainerDefault from "./common/RowCountWarningContainer";

import NonStickyComponent from "./NonSticky";
import StickyComponent from "./Sticky";


/*
    flex: 1 1 auto, assuming that table would be used full-stretch mostly
*/
const wrapperClass = css`
    min-height: 0;
    flex: 1 1 auto;

    td, th {
        overflow: hidden;
        text-overflow: ellipsis;
    }

    thead, tfoot {
        white-space: nowrap;
    }

    * {
        box-sizing: border-box;
    }
`;

/*
    dataRef is to call Data methods from outside( Data.scrollTo(), etc. ).
    As it is not dom-related, I decided to avoid forwardRef
*/
const Table = ({
    columns,
    totals,
    getRowData,
    getRowKey,
    getRowExtraProps,
    rowCount,
    estimatedRowHeight,
    overscanRowsCount,
    rowCountWarningsTable,
    headless,
    RowCountWarningContainer,
    dataRef,
    useStickyIfPossible,

    ...props
}) => {

    const scrollContainerRef = useRef();
    const tbodyRef = useRef();
    const isMountedRef = useRef();
    const finalDataRef = useRef();

    if( !finalDataRef.current ){
        finalDataRef.current = new VirtualTableDataStore({
            overscanRowsCount,
            columns,
            totals,
            totalRows: rowCount,
            rowDataGetter: getRowData,
            rowKeyGetter: getRowKey,
            estimatedRowHeight: estimatedRowHeight,
            headlessMode: headless,
            getRowsContainerNode: () => tbodyRef.current,
            getScrollContainerNode: () => scrollContainerRef.current
        });
    }

    useEffect(() => {
        if( dataRef ){
            dataRef.current = finalDataRef.current;
        }
        if( isMountedRef.current ){
            finalDataRef.current
                .setHeadlessMode( headless )
                .setRowDataGetter( getRowData )
                .setRowKeyGetter( getRowKey )
                .setOverscanRowsCount( overscanRowsCount )
                .setTotals( totals )
                .setColumns( columns )
                .setTotalRows( rowCount )
                .setEstimatedRowHeight( estimatedRowHeight );
        }
    });

    useEffect(() => {
        isMountedRef.current = true;
        return () => {
            finalDataRef.current.destructor();
        };
    }, []);

    /*
        Only cells inside thead/tfoot can be sticky.
        If thead/tfoot are hidden - we can easily render lighter StickyComponent to avoid extra wrappers
    */
    const ComponentVariant = ( headless && !totals ) || ( useStickyIfPossible && isPositionStickySupported() ) ? StickyComponent : NonStickyComponent;

    return (
        <Context.Provider value={finalDataRef.current}>
            { rowCount > 0 ? (
                <ComponentVariant
                    className={wrapperClass}
                    scrollContainerRef={scrollContainerRef}
                    tbodyRef={tbodyRef}
                    {...props}
                />
            ) : rowCountWarningsTable ? (
                <RowCountWarningContainer>
                    {rowCountWarningsTable[rowCount]}
                </RowCountWarningContainer>
            ) : null }
        </Context.Provider>
    );
}

Table.propTypes = {
    columns: PropTypes.arrayOf(
        PropTypes.shape({
            /* unique key for column */
            dataKey: PropTypes.string.isRequired,

            /* 
                If rowData is available, cellData goes through flow, where each fn is optional: transformCellData(getCellData(rowData,rowIndex),rowData, column, rowIndex)
                If not, it goes through flow: getEmptyCellData(rowIndex, column).
            */
            getCellData: PropTypes.func,
            getEmptyCellData: PropTypes.func,
            transformCellData: PropTypes.func,

            visibility: PropTypes.oneOf([ "visible", "hidden" ]),
            sort: PropTypes.oneOf([ "locale", "numeric" ]),

            /*
                column props, affecting colgroup > col tags
            */
            background: PropTypes.string,
            border: PropTypes.string,
            width: PropTypes.oneOfType([ PropTypes.number, PropTypes.string ])
        })
    ).isRequired,
    getRowData: PropTypes.func.isRequired,

    totals: PropTypes.objectOf(
        /* array may contain: "sum", "average", "count". */
        PropTypes.array
    ),
    
    useStickyIfPossible: PropTypes.bool,
    headless: PropTypes.bool,
    className: PropTypes.string,
    rowCount: PropTypes.number,
    getRowKey: PropTypes.func,
    estimatedRowHeight: PropTypes.number,
    getRowExtraProps: PropTypes.func,
    overscanRowsCount: PropTypes.number,

    HeaderRowComponent: PropTypes.any,
    RowComponent: PropTypes.any,
    CellComponent: PropTypes.any,

    RowCountWarningContainer: PropTypes.any,
    rowCountWarningsTable: PropTypes.object,

    /* Determines, if table-layout: fixed is applied to main table */
    fixedLayout: PropTypes.bool
};

Table.defaultProps = {
    rowCount: 0,
    estimatedRowHeight: 20,
    overscanRowsCount: 4,
    fixedLayout: false,
    headless: false,

    /*
        For 90% non-reactive solutions, which only provide new getRowData when data is changed, memo is ok.
        If RowComponent should be wrapped my mobx observer - non-memo version should be imported.
        memo(observer(RowComponentDefault)) will do the trick.
    */
    RowComponent: memo( RowComponentDefault ),
    CellComponent: CellComponentDefault,
    RowCountWarningContainer: RowCountWarningContainerDefault,
};

export default memo( Table );