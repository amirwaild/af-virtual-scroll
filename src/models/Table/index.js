import debounce from "lodash/debounce";
import subtract from "lodash/subtract";
import addSetters from "../../utils/addSetters";
import List from "../List";

const REFRESH_SORT_DEBOUNCE_INTERVAL = 500;

const getRowDataInitial = () => {
    throw new Error( "getRowData must be provided for table" );
};

const fillOrderedRowsArray = ( arr, startIndex, endIndex ) => {
    while( startIndex < endIndex ){
        arr[ startIndex ] = startIndex++;
    }
};

const L = new Intl.Collator();

const getSorter = ( getRowData, fieldName, method, getCellData, directionSign ) => {
    const fn = method === "locale" ? L.compare : subtract;
    const defaultValue = method === "locale" ? "" : 0;

    if( getCellData ){
        return ( a, b, i ) => {
            a = getRowData( a );
            a = a ? getCellData( a, i ) : defaultValue;
            b = getRowData( b );
            b = b ? getCellData( b, i ) : defaultValue;
            return fn( a, b ) * directionSign;
        };
    }
    
    return ( a, b ) => {
        a = getRowData( a );
        a = a ? a[ fieldName ] : defaultValue;
        b = getRowData( b );
        b = b ? b[ fieldName ] : defaultValue;
        return fn( a, b ) * directionSign;
    };
};

const calculateSum = ( totalRows, dataKey, getRowData, getCellData ) => {
    let res = 0;
    for( let i = 0, rowData, cellData; i < totalRows; i++ ){
        rowData = getRowData( i );
        cellData = getCellData ? getCellData( rowData, i ) : rowData[ dataKey ];
        if( cellData ){
            res += cellData;
        }
    }
    return res;
};

/*
    We could use simple object literal,
    but constructors with stable-order this initialization enforce "hidden-classes" v8 optimization
*/
class TotalsCachePart {
    count = 0;
    sum = 0;
    average = 0.0;
};

class Table extends List {

    sortColumnIndex = -1;
    sortDirectionSign = 1;

    scrollLeft = 0;
    tbodyColumnWidths = null;
    orderedRows = [];

    /*
        We do not want to recalculate totals too often, so caching them in object by column dataKey
    */
    totalsCache = Object.create( null );


    /*
        TODO:
            make this call throttled
    */
    refreshTotalsForColumn( dataKey ){
        const curTotals = this.totals[ dataKey ];
        if( curTotals ){
            let curCachePart = this.totalsCache[ dataKey ];

            if( !curCachePart ){
                curCachePart = this.totalsCache[ dataKey ] = new TotalsCachePart();
            }
            
            for( let j = 0, totalType, oldVal, newVal; j < curTotals.length; j++ ){
                totalType = curTotals[ j ];
                oldVal = curCachePart[ totalType ];
                switch( curTotals[ j ] ){
                    case "count":
                        newVal = this.totalRows;
                        break;
                    case "sum":
                        newVal = calculateSum( this.totalRows, dataKey, this.rowDataGetter );
                        break;
                    case "average":
                        /* Todo: optimize, if we already calculated sum */
                        newVal = calculateSum( this.totalRows, dataKey, this.rowDataGetter ) / this.totalRows;
                        break;
                    default:
                        if( process.env.NODE_ENV !== "production" ){
                            console.error( `Unknown totals type: ${curTotals[ j ]}` );
                        }
                }
                if( oldVal !== newVal ){
                    curCachePart[ totalType ] = newVal;
                    this.emit( "totals-calculated" );
                }
            }
        }
        else if( process.env.NODE_ENV !== "production" ){
            console.log( `Asked to recalculate totals for: ${dataKey}; doing nothing;` );
        }
        return this;
    }

    refreshTotals = debounce(() => {
        for( let j = 0, dataKey; j < this.columns.length; j++ ){
            dataKey = this.columns[ j ].dataKey;
            this.refreshTotalsForColumn( dataKey );
        }
    }, 100 );

    setSortParams( colIndex, sortDirectionSign ){
        if( this.sortColumnIndex !== colIndex || sortDirectionSign !== this.sortDirectionSign ){
            this.sortColumnIndex = colIndex;
            this.sortDirectionSign = sortDirectionSign;
            this.emit( "sort-params-changed" );
        }
    }

    refreshSorting = debounce(() => {
        if( this.sortColumnIndex > -1 && this.totalRows > 0 ){
            const { sort, dataKey, getCellData } = this.columns[ this.sortColumnIndex ];
            if( sort ){
                const sorter = getSorter( this.rowDataGetter, dataKey, sort, getCellData, this.sortDirectionSign );
                this.orderedRows.sort( sorter );
                this.emit( "rows-order-changed" );
            }
        }
    }, REFRESH_SORT_DEBOUNCE_INTERVAL );

    refreshRowsOrder( prevTotalRows ){
        if( this.totalRows > 0 ){
            this.orderedRows.length = this.totalRows;
            fillOrderedRowsArray( this.orderedRows, this.totalRows > prevTotalRows ? prevTotalRows : 0, this.totalRows );
        }
        else{
            this.orderedRows.length = 0;
        }
        return this;
    }

    resetColumnWidthsCache(){
        this.tbodyColumnWidths = new Uint32Array( this.columns.length );
    }

    constructor( params ){
        super( params );

        this
            .on( "columns-changed", this.resetColumnWidthsCache, this )
            .on( "columns-changed", this.refreshTotals )
            .on( "total-rows-changed", this.refreshRowsOrder, this )
            .on( "total-rows-changed", this.refreshSorting )
            .on( "total-rows-changed", this.refreshTotals )
            .on( "sort-params-changed", this.refreshSorting )
            .on( "columns-changed", this.refreshSorting )
            .on( "row-data-getter-changed", this.refreshSorting )
            .on( "row-data-getter-changed", this.refreshTotals )
            .on( "rows-order-changed", this.resetMeasurementsCache, this )
            .on( "rows-order-changed", () => this.scrollToRow( 0 ) )
            .on( "totals-changed", this.refreshTotals )
        
            .setColumns( params.columns || [] )
            .setTotals( params.totals )
            .setHeadlessMode( !!params.headlessMode )
            .setRowDataGetter( params.rowDataGetter || getRowDataInitial )
            .refreshRowsOrder( 0 );
    }

    destructor(){
        this.refreshSorting.cancel();
        this.refreshTotals.cancel();
        super.destructor();
    }
}

addSetters( Table.prototype, [
    "columns",
    "totals",
    "scrollLeft",
    "rowDataGetter",
    "headlessMode"
]);

export default Table;