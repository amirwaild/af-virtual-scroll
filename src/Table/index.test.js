import React from "react";
import { create, act } from "react-test-renderer";
import Table from "./index";

describe( "Table component renders ok", () => {

    test.each([ undefined, -2, -1, 0, 1, 2 ].map( v => [ v ] ))( "Performs initial render with rowCount: %p", rowCount => {
        let component;
        act(() => {
            component = create(
                <Table
                    columns={[]}
                    getRowData={() => ({ a: 1 })}
                    rowCount={rowCount}
                />
            );
        });
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});