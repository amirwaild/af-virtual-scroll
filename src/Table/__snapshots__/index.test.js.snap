// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component renders ok Performs initial render with rowCount: -1 1`] = `null`;

exports[`Table component renders ok Performs initial render with rowCount: -2 1`] = `null`;

exports[`Table component renders ok Performs initial render with rowCount: 0 1`] = `null`;

exports[`Table component renders ok Performs initial render with rowCount: 1 1`] = `
<div
  className="afvscr-nonst afvscr-table-wrapper"
>
  <table
    aria-colcount={0}
    className="afvscr-nonst-subtable"
    style={
      Object {
        "marginLeft": -0,
        "width": 0,
      }
    }
  >
    <colgroup />
    <thead
      onClick={[Function]}
    >
      <tr />
    </thead>
  </table>
  <div
    className="afvscr-scroll-container"
    onScroll={[Function]}
    tabIndex="0"
  >
    <div
      aria-hidden="true"
      className="afvscr-height-provider"
      style={
        Object {
          "height": 16,
        }
      }
    />
    <table
      aria-colcount={0}
      aria-rowcount={1}
      style={
        Object {
          "minWidth": "100%",
          "tableLayout": "auto",
        }
      }
    >
      <colgroup />
      <thead
        className="afvscr-hdnwrp"
        onClick={[Function]}
      >
        <tr />
      </thead>
      <tbody
        aria-hidden="true"
        className="afvscr-scroller"
      >
        <tr
          aria-hidden="true"
          className="afvscr-scroller"
          style={
            Object {
              "height": 0,
            }
          }
        />
      </tbody>
      <tbody>
        <tr
          aria-rowindex={1}
        />
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table component renders ok Performs initial render with rowCount: 2 1`] = `
<div
  className="afvscr-nonst afvscr-table-wrapper"
>
  <table
    aria-colcount={0}
    className="afvscr-nonst-subtable"
    style={
      Object {
        "marginLeft": -0,
        "width": 0,
      }
    }
  >
    <colgroup />
    <thead
      onClick={[Function]}
    >
      <tr />
    </thead>
  </table>
  <div
    className="afvscr-scroll-container"
    onScroll={[Function]}
    tabIndex="0"
  >
    <div
      aria-hidden="true"
      className="afvscr-height-provider"
      style={
        Object {
          "height": 32,
        }
      }
    />
    <table
      aria-colcount={0}
      aria-rowcount={2}
      style={
        Object {
          "minWidth": "100%",
          "tableLayout": "auto",
        }
      }
    >
      <colgroup />
      <thead
        className="afvscr-hdnwrp"
        onClick={[Function]}
      >
        <tr />
      </thead>
      <tbody
        aria-hidden="true"
        className="afvscr-scroller"
      >
        <tr
          aria-hidden="true"
          className="afvscr-scroller"
          style={
            Object {
              "height": 0,
            }
          }
        />
      </tbody>
      <tbody>
        <tr
          aria-rowindex={1}
        />
        <tr
          aria-rowindex={2}
        />
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table component renders ok Performs initial render with rowCount: undefined 1`] = `null`;
